### PR TITLE
Disable optional rules by default on the NuGet package scripts

### PR DESCRIPTION
**Pull request by pjanotti**
## Why

What these rules validate can't happen when using the NuGet.

## What

Set `OTEL_DOTNET_AUTO_RULE_ENGINE_ENABLED` to `false` on the script shipped by the NuGet package.

## Tests

Locally run StartupHook.Tests to validate cea5e15389c3735f77a0c5412eeee5ec58b25472

## Checklist

~~- [ ] `CHANGELOG.md` is updated.~~
~~- [ ] Documentation is updated.~~
~~- [ ] New features are covered by tests.~~

### CODE REVIEW

File: src/OpenTelemetry.AutoInstrumentation.StartupHook/RulesEngine/RuleEngine.cs [hunk 1]
```diff
@@ -28,23 +28,8 @@ internal class RuleEngine
...
-
-    internal RuleEngine()
-    {
-    }
-
-    // This constructor is used for test purpose.
-    internal RuleEngine(List<Rule> rules)
-    {
-        _otherRules = rules;
-    }
```
Comment (nrcventura) [inline src/OpenTelemetry.AutoInstrumentation.StartupHook/RulesEngine/RuleEngine.cs:47]: Did you need to change the way the rules are injected? You probably could have kept the existing injection pattern but added a new private field for the `_optionalTests` and made `_otherRules` static. Then you could update the .ctors to do something like the following.
```
internal RuleEngine() : this(_otherRules) {}

//This constructor is used for test purpose
internal RuleEngine(List<Rule> optionalRules)
{
    _optionalRules = optionalRules;
}
```

Comment (pjanotti) [reply src/OpenTelemetry.AutoInstrumentation.StartupHook/RulesEngine/RuleEngine.cs:47]: It is not required to change the way that the injection happens as long as the list of optional rules is only created when it is going to be used. Let me play with it.

Comment (pjanotti) [reply src/OpenTelemetry.AutoInstrumentation.StartupHook/RulesEngine/RuleEngine.cs]: Returned to prior injection pattern.

File: src/OpenTelemetry.AutoInstrumentation.StartupHook/RulesEngine/RuleEngine.cs [hunk 2]
```diff
@@ -65,8 +59,16 @@ internal bool ValidateRules()
             return result;
         }
 
+        var optionalRules = _optionalRules ?? new()
+        {
+            new OpenTelemetrySdkMinimumVersionRule(),
+            new DiagnosticSourceRule(),
+            new InstrumentationAssemblyRule(),
+            new NativeProfilerDiagnosticsRule()
```
Comment (nrcventura) [inline src/OpenTelemetry.AutoInstrumentation.StartupHook/RulesEngine/RuleEngine.cs:67]: Why do we need to defer the creation of these rules?

Comment (pjanotti) [reply src/OpenTelemetry.AutoInstrumentation.StartupHook/RulesEngine/RuleEngine.cs:67]: If we create them before knowing if they will be used or not we can load assemblies, make allocations, etc... that end up being not used. Delaying this creation until we are sure that they will be used avoid that, it becomes pay-for-play.

**Review by rajkumar-rangaraj — APPROVED** [review]

File: src/OpenTelemetry.AutoInstrumentation.StartupHook/RulesEngine/RuleEngine.cs (hunk 2 above)
Comment (pjanotti) [reply src/OpenTelemetry.AutoInstrumentation.StartupHook/RulesEngine/RuleEngine.cs:67]: Check lines 56-60 above.

Comment (pjanotti) [reply src/OpenTelemetry.AutoInstrumentation.StartupHook/RulesEngine/RuleEngine.cs]: > So just the act of creating these types causes other dependencies to load?

It depends on what the constructors of each type do. The current implementations don't do that, but, there is always the possibility of some later change creating or modifying a rule type that does that.

I like the idea of a static method to make clear that we are creating the default rules. Updating the code...

Comment (pjanotti) [reply src/OpenTelemetry.AutoInstrumentation.StartupHook/RulesEngine/RuleEngine.cs]: 5efcf1e04ddf61b405022472826f386113d35495

**Review by nrcventura — APPROVED** [review]

**Pull request merged**